test/e2e_node/image_list.go

```diff
@@ -151,27 +152,39 @@ func PrePullAllImages() error {
 	}
 	images := framework.ImageWhiteList.List()
 	klog.V(4).Infof("Pre-pulling images with %s %+v", puller.Name(), images)
+
+	var wg sync.WaitGroup
+	var once sync.Once
+	wg.Add(len(images))
 	for _, image := range images {
-		var (
-			err    error
-			output []byte
-		)
-		for i := 0; i < maxImagePullRetries; i++ {
-			if i > 0 {
-				time.Sleep(imagePullRetryDelay)
+		go func(image string) {
+			defer wg.Done()
+
+			var (
+				pullErr error
+				output  []byte
+			)
+			for i := 0; i < maxImagePullRetries; i++ {
+				if i > 0 {
+					time.Sleep(imagePullRetryDelay)
+				}
+				if output, pullErr = puller.Pull(image); pullErr == nil {
+					break
+				}
+				klog.Warningf("Failed to pull %s as user %q, retrying in %s (%d of %d): %v",
+					image, usr.Username, imagePullRetryDelay.String(), i+1, maxImagePullRetries, pullErr)
 			}
-			if output, err = puller.Pull(image); err == nil {
-				break
+			if pullErr != nil {
```

Comment:
One behavior from the serialized approach was that if we failed to pull any single image, we would immediately fail the whole function. Thoughts on if we want to retain that behavior in the parallel approach? Right now, we wait for all of the images to attempt a pull before returning any type of error value.

Reply:
The **puller.Pull** doesn't have any param like **ctx** to cancel the process(neither exec.Command nor ImageService interface), so we must wait once pull returned(If we exit **PrePullAllImages** without wait sub-goroutine exited, there is goroutine-leak happened).

In other words, we can immediately fail the pull process when every pull-retry started.

Reply:
Just to confirm - you're saying that we can't cancel the current pull, BUT we can prevent it from retrying if it fails (if another image pull has already failed)? I'm comfortable with waiting until the possible retry to fail (and agree with you there isn't another option). I still think that's preferable to letting all of the attempted pulls go through the full number of retries.

Reply:
Yes, we can prevent it from retrying if any single image pull failed.